src/Pos/Client/Txp/Util.hs

```diff
@@ -206,60 +260,184 @@ prepareInpsOuts utxo addrs outputs = do
                 case mNextOut of
-                    Nothing -> throwError "Not enough money to send!"
+                    Nothing -> throwTxError $
+                        sformat ("Not enough money to send (need "%build%" coins more)") moneyLeft
                     Just inp@(_, (TxOutAux (TxOut {..}) _)) -> do
                         _1 .= unsafeSubCoin moneyLeft (min txOutValue moneyLeft)
                         _2 %= tail
                         pickInputs (inp : inps)
     formTxInputs (inp, TxOutAux TxOut{..} _) = (txOutAddress, inp)
 
--- | Common use case of 'prepaseInpsOuts' - with single source address
-prepareInpOuts :: MonadError Text m => Utxo -> Address -> TxOutputs -> m (TxInputs, TxOutputs)
-prepareInpOuts utxo addr outputs =
-    prepareInpsOuts utxo (one addr) outputs <&>
-    _1 . traversed %~ snd
+-- | Returns set of tx outputs including change output (if it's necessary)
+mkOutputsWithRem
+    :: MonadAddresses m
+    => AddrData m
+    -> TxRaw
+    -> m TxOutputs
+mkOutputsWithRem addrData TxRaw {..}
+    | trRemaining == mkCoin 0 = pure trOutputs
+    | otherwise = do
+          changeAddr <- getNewAddress addrData
+          pure $ (TxOutAux (TxOut changeAddr trRemaining) []) :| toList trOutputs
 
 -- | Make a multi-transaction using given secret key and info for outputs.
 -- Currently used for HD wallets only, thus `HDAddressPayload` is required
-createMTx :: Utxo -> NonEmpty (SafeSigner, Address) -> TxOutputs -> Either Text TxAux
-createMTx utxo hwdSigners outputs =
-    let addrs = map snd hwdSigners
-    in  uncurry (makeMPubKeyTx getSigner) <$>
-        prepareInpsOuts utxo addrs outputs
-  where
-    signers = HM.fromList . toList $ map (swap . second AddressIA) hwdSigners
-    getSigner addr =
-        fromMaybe (error "Requested signer for unknown address") $
-        HM.lookup (AddressIA addr) signers
+createMTx
+    :: TxCreateMode ctx m
+    => Utxo
+    -> NonEmpty (SafeSigner, Address)
+    -> TxOutputs
+    -> AddrData m
+    -> m (Either TxError TxAux)
+createMTx utxo hdwSigners outputs addrData = runExceptT $ do
+    txRaw@TxRaw {..} <- prepareTxWithFee utxo outputs
+    outputsWithRem <- lift $ mkOutputsWithRem addrData txRaw
+    properOutputs <- overrideTxDistrBoot outputsWithRem
+    pure $ makeMPubKeyTxAddrs hdwSigners trInputs properOutputs
 
 -- | Make a multi-transaction using given secret key and info for
 -- outputs.
-createTx ::
-       TxCreateMode ctx m
+createTx
+    :: TxCreateMode ctx m
     => Utxo
     -> SafeSigner
     -> TxOutputs
-    -> m (Either Text TxAux)
-createTx utxo ss outputs =
-    runExceptT $ do
-        properOutputs <- overrideTxDistrBoot outputs
-        uncurry (makePubKeyTx ss) <$>
-            prepareInpOuts
-                utxo
-                (makePubKeyAddress $ safeToPublic ss)
-                properOutputs
+    -> AddrData m
+    -> m (Either TxError TxAux)
+createTx utxo ss outputs addrData = runExceptT $ do
```

Comment:
`createTx`, `createMTx` and some functions below got almost identical, it's worth extracting some logic